File: JasSwift/Sources/Algorithms/PatternAlongPath.swift
import Foundation

// Pattern brush: side artwork tile repeated along the stroke path.
// Port of jas_dioxus/src/algorithms/pattern_along_path.rs (BRUSHES.md
// §Brush types > Pattern). Reuses flattenArtPath + pointAtArcLength.
// Phase 1: SIDE tile only (corner tiles deferred), polygon artwork.

public struct PatternBrush: Equatable {
    public let tileWidth: Double
    public let tileHeight: Double
    public let side: [[[Double]]]   // side-tile polygons of [x, y] pairs
    public let scale: Double        // percent
    public let spacing: Double      // percent of tile width
    public let flipAcross: Bool
    public let flipAlong: Bool
    public let strokeWeight: Double // pt

    public init(tileWidth: Double, tileHeight: Double, side: [[[Double]]],
                scale: Double, spacing: Double, flipAcross: Bool, flipAlong: Bool,
                strokeWeight: Double) {
        self.tileWidth = tileWidth
        self.tileHeight = tileHeight
        self.side = side
        self.scale = scale
        self.spacing = spacing
        self.flipAcross = flipAcross
        self.flipAlong = flipAlong
        self.strokeWeight = strokeWeight
    }
}

/// Tile `brush.side` along the stroke `commands`. Returns one warped polygon
/// per (tile placement × side polygon); empty for degenerate input.
public func patternAlongPath(_ commands: [PathCommand], _ brush: PatternBrush) -> [[[Double]]] {
    guard brush.tileWidth > 0, brush.tileHeight > 0 else { return [] }
    let pts = flattenArtPath(commands)
    if pts.count < 2 { return [] }
    var cum = [Double](repeating: 0, count: pts.count)
    for i in 1..<pts.count {
        let dx = pts[i].0 - pts[i - 1].0
        let dy = pts[i].1 - pts[i - 1].1
        cum[i] = cum[i - 1] + (dx * dx + dy * dy).squareRoot()
    }
    let total = cum[pts.count - 1]
    if total <= 0 { return [] }
    let ribbon = (brush.scale / 100.0) * brush.strokeWeight
    let tileW = ribbon * (brush.tileWidth / brush.tileHeight)
    if tileW <= 0 { return [] }
    let gap = tileW * (brush.spacing / 100.0)
    let step = tileW + gap
    if step <= 0 { return [] }
    let n = max(Int((total / step).rounded(.down)), 1)

    var out: [[[Double]]] = []
    for i in 0..<n {
        let start = Double(i) * step
        for poly in brush.side {
            var warped: [[Double]] = []
            for pair in poly {
                let ax = pair[0], ay = pair[1]
                var u = min(max(ax / brush.tileWidth, 0.0), 1.0)
                if brush.flipAlong { u = 1.0 - u }
                let s = start + u * tileW
                let (px, py, tan) = pointAtArcLength(pts, cum, total, s)
                var off = (ay - brush.tileHeight / 2.0) / brush.tileHeight * ribbon
                if brush.flipAcross { off = -off }
                let nx = -sin(tan), ny = cos(tan)
                warped.append([px + nx * off, py + ny * off])
            }
            out.append(warped)
        }
    }
    return out
}

File: jas/algorithms/pattern_along_path.py
"""Pattern brush: the side artwork tile repeated along the stroke path.

Port of jas_dioxus/src/algorithms/pattern_along_path.rs (BRUSHES.md §Brush
types > Pattern). Reuses art_along_path's _flatten + _point_at_arclength.
Phase 1: SIDE tile only (corner tiles deferred), polygon artwork.
"""

import math
from dataclasses import dataclass

from algorithms.art_along_path import _flatten, _point_at_arclength


@dataclass
class PatternBrush:
    tile_width: float
    tile_height: float
    side: list       # list of side-tile polygons, each a list of (x, y)
    scale: float     # percent
    spacing: float   # percent of tile width
    flip_across: bool
    flip_along: bool
    stroke_weight: float  # pt


def pattern_along_path(commands, brush: "PatternBrush"):
    """Tile ``brush.side`` along ``commands``. Returns one warped polygon
    per (tile placement x side polygon); [] for degenerate input."""
    if brush.tile_width <= 0 or brush.tile_height <= 0:
        return []
    pts = _flatten(commands)
    if len(pts) < 2:
        return []
    cum = [0.0] * len(pts)
    for i in range(1, len(pts)):
        cum[i] = cum[i - 1] + math.hypot(pts[i][0] - pts[i - 1][0],
                                         pts[i][1] - pts[i - 1][1])
    total = cum[-1]
    if total <= 0:
        return []
    ribbon = (brush.scale / 100.0) * brush.stroke_weight
    tile_w = ribbon * (brush.tile_width / brush.tile_height)
    if tile_w <= 0:
        return []
    gap = tile_w * (brush.spacing / 100.0)
    step = tile_w + gap
    if step <= 0:
        return []
    n = max(int(total // step), 1)
    out = []
    for i in range(n):
        start = i * step
        for poly in brush.side:
            warped = []
            for ax, ay in poly:
                u = min(max(ax / brush.tile_width, 0.0), 1.0)
                if brush.flip_along:
                    u = 1.0 - u
                s = start + u * tile_w
                px, py, tan = _point_at_arclength(pts, cum, total, s)
                off = (ay - brush.tile_height / 2.0) / brush.tile_height * ribbon
                if brush.flip_across:
                    off = -off
                nx = -math.sin(tan)
                ny = math.cos(tan)
                warped.append((px + nx * off, py + ny * off))
            out.append(warped)
    return out
